### PR TITLE
ad/mps

### PR DESCRIPTION
**Pull request by annamariadziubyna**
- correct _holes 
- add new ways of working with gates in MPS2
- add tests

### CODE REVIEW

**Comment by lpawela** [comment]
@annamariadziubyna please provide a short summary of changes this PR introduces and also resolve any conflicts.

**Review by bartekGardas — CHANGES_REQUESTED** [review]
1. _apply_layer_of_gates takes way too many arguments, some of them are redundant, e.g., L (as it is already included in ig). In cases like this, we rather use MPScontrol.

2. What Dl1⊗Dl2 does, why not?

**Review by bartekGardas — CHANGES_REQUESTED** [review]
1. Instead of using setdiff(i+1:last(nbrs), nbrs), please define correct _holes from utils.jl.

File: src/spectrum.jl [hunk 1]
```diff
@@ -181,14 +181,20 @@ function multiply_purifications(χ::AbstractMPS, ϕ::AbstractMPS, L::Int)
         A1 = χ[i]
         A2 = ϕ[i]
         
-        @cast B[Dl1⊗Dl2, d, Dr1⊗Dr2] := A1[Dl1, d, Dr1] * A2[Dl2, d, Dr2]
+        @cast B[(Dl1, Dl2), d, (Dr1, Dr2)] := A1[Dl1, d, Dr1] * A2[Dl2, d, Dr2]
```
Comment (lpawela) [inline src/spectrum.jl]: Please follow the convention on index naming - greek vs latin letters

File: src/spectrum.jl [hunk 2]
```diff
@@ -213,15 +216,84 @@ function MPS(ig::MetaGraph, control::MPSControl)
...
+    dβ = get_prop(ig, :dβ)
+    β = get_prop(ig, :β)
+    rank = get_prop(ig, :rank)
+
+    @info "Preparing Hadamard state as MPS"
+    ρ = HadamardMPS(rank)
+    is_right = true
+    @info "Sweeping through β and σ" dβ
+
+    if control.type == "log"
```
Comment (lpawela) [inline src/spectrum.jl]: Too much code repetition

File: test/searchMPS.jl [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+type1 = "log"
+type2 = "lin"
+β = [get_prop(ig, :β)]
+dβ = [get_prop(ig, :dβ)]
+control1 = MPSControl(D, var_ϵ, sweeps, β, dβ, type1) 
+control2 = MPSControl(D, var_ϵ, sweeps, β, dβ, type2) 
+
+states = all_states(get_prop(ig, :rank))
+ϱ = gibbs_tensor(ig)
+@test sum(ϱ) ≈ 1
```
Comment (lpawela) [inline test/searchMPS.jl:28]: This is the only test in this file and it is trivial. Please ad some more tests.

File: Project.toml [hunk 4]
```diff
@@ -9,9 +9,13 @@ CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
 DocStringExtensions = "ffbed154-4ef7-542d-bbb7-c09d3a79fcae"
 LightGraphs = "093fc24a-ae57-5d10-9952-331d41423f4d"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Logging = "56ddb016-857b-54e1-b83d-db4d58db5568"
```
Comment (lpawela) [inline Project.toml]: Package that are listed under `[extras]` should not be here.

**Review by lpawela — CHANGES_REQUESTED** [review]

**Review by lpawela — APPROVED** [review]

**Comment by lpawela** [comment]
@annamariadziubyna please resolve these conflits and then this can be merged

**Comment by lpawela** [comment]
@annamariadziubyna please fix test coverage

**Review by bartekGardas — APPROVED** [review]
Let's merge now.

**Pull request merged**